qutip/scattering.py

```diff
@@ -0,0 +1,398 @@
+Pulses from Quantum Optical Systems (2017, arXiv:1710.02875).
+"""
+
+# Author:  Ben Bartlett
+# Contact: benbartlett@stanford.edu
+
+__all__ = ['temporal_basis_vector',
+           'temporal_scattered_state',
+           'scattering_probability']
+
+import numpy as np
+from itertools import product, combinations_with_replacement
+from qutip import propagator, Options, basis, tensor, zero_ket, Qobj
+
+
+class Evolver:
+    """
+    A caching class which takes a Hamiltonian and a list of times to calculate
+    and memoize propagators for the system between any two times as demanded.
+
+    Parameters
+    ----------
+    H : Qobj or list
+        system-waveguide(s) Hamiltonian or effective Hamiltonian in Qobj or
+        list-callback format. If construct_effective_hamiltonian is not
+        specified, an effective Hamiltonian is constructed from H and c_ops
+    times: list-like
+        list of times to evaluate propagators over
+    options: qutip.Options
+        solver options to use when computing propagators
+
+    Attributes
+    ----------
+    H: Qobj
+        system-waveguide(s) Hamiltonian, may be time-dependent
+    tlist: list-like
+        list of times to evaluate propagators over
+    propagators: (dict of float: (dict of float: qutip.Qobj))
+        dictionary of dictionaries of propagator objects with keys of
+        evaluation times, e.g. propagators[t2][t1] returns U[t2,t1]
+
+    """
+
+    def __init__(self, H, tlist, options = None):
+        self.H = H
+        self.tlist = tlist
+        if options is not None:
+            self.options = options
+        else:
+            self.options = Options(nsteps = 10000, normalize_output = False)
+        # Make a blank nested dictionary to store propagators
+        self.propagators = dict.fromkeys(tlist)
+        for t in tlist:
+            self.propagators[t] = dict.fromkeys(tlist)
+
+    def prop(self, tf, ti):
+        """Compute U[t2,t1] where t2 > t1 or return the cached operator
+
+        Parameters
+        ----------
+        tf: float
+            final time to compute the propagator U[tf,ti] for; the propagator
+            is computed to the nearest time in self.times
+        ti: float
+            initial time to compute the propagator U[tf,ti] for
+
+        Returns
+        -------
+        propagator: Qobj
+            the propagation operator
+        """
+        left, right = np.searchsorted(self.tlist, [ti, tf], side = 'left')
+        t1, t2 = self.tlist[left], self.tlist[right]
+        if self.propagators[t2][t1] is None:
+            self.propagators[t2][t1] = propagator(self.H, [t1, t2],
+                                                  options = self.options,
+                                                  unitary_mode = 'single')
+            # Something is still broken about batch unitary mode (see #807)
+        return self.propagators[t2][t1]
+
+    def count_filled(self):
```

Comment:
Function not covered by tests.

Reply:
I'd actually rather just remove the `count_filled()` method; it's just a debugging remnant from when I was developing this module. If you can think of a reason it should remain, let me know.